### ISSUE TITLE
[Feature] Eager element-wise ops with broadcasting (CPU + CUDA).

### ISSUE THREAD

**Issue by Hayden727** [Hayden727/ctorch#3]
<h2 data-heading="1. Background">1. Background</h2>
<p>With Tensor + dispatch in place, the next milestone is the first real ops. Element-wise<br>
binary and unary operations are the smallest set that exercises the full path:<br>
broadcasting, dispatch, both backends, type promotion, and the parity-test harness<br>
against PyTorch. They're also the operations every higher-level component (autograd,<br>
nn, dynamo) implicitly relies on.</p>
<p>This issue ships those ops on <strong>both</strong> CPU and CUDA in one go, because numerical parity<br>
between backends is a property we want to lock in from day 1.</p>
<hr>
<h2 data-heading="2. Requirements">2. Requirements</h2>
<h3 data-heading="2.1 Functional">2.1 Functional</h3>
<ul>
<li><strong>F1.</strong> Binary ops with broadcasting: <code>add</code>, <code>sub</code>, <code>mul</code>, <code>div</code>, plus in-place<br>
variants (<code>add_</code>, <code>sub_</code>, <code>mul_</code>, <code>div_</code>).</li>
<li><strong>F2.</strong> Unary ops: <code>neg</code>, <code>abs</code>, <code>exp</code>, <code>log</code>, <code>sqrt</code>, <code>relu</code>, <code>sigmoid</code>, <code>tanh</code>.</li>
<li><strong>F3.</strong> Operator overloads on <code>Tensor</code>: <code>+</code>, <code>-</code>, <code>*</code>, <code>/</code>, unary <code>-</code>.</li>
<li><strong>F4.</strong> Broadcasting follows NumPy/PyTorch rules; no copy of inputs is required when<br>
shapes are compatible.</li>
<li><strong>F5.</strong> Type promotion (<code>promote_types</code>) matches PyTorch's table for the supported<br>
dtypes (<code>float32</code>, <code>float64</code>, <code>int32</code>, <code>int64</code>, <code>bool</code>).</li>
<li><strong>F6.</strong> Output dtype is determined by promotion; output device matches the (single)<br>
input device — cross-device ops throw a clear error.</li>
<li><strong>F7.</strong> Both CPU and CUDA implementations are registered into the dispatch table;<br>
<code>t.add(u)</code> selects the right kernel from <code>t.device()</code>.</li>
</ul>
<h3 data-heading="2.2 Non-functional">2.2 Non-functional</h3>
<ul>
<li><strong>N1.</strong> Numerical parity vs PyTorch reference: fp32 within <code>1e-5</code> (rel) for non-trans<br>
cendental ops, <code>1e-4</code> for <code>exp</code>/<code>log</code>/<code>sqrt</code>/<code>sigmoid</code>/<code>tanh</code>; fp64 within <code>1e-12</code>.</li>
<li><strong>N2.</strong> CUDA <code>add</code> on a 1M-element fp32 tensor lands within <strong>10 %</strong> of a hand-written<br>
reference kernel that does the same global-memory traffic — proves no overhead from<br>
the dispatcher or the indexer.</li>
<li><strong>N3.</strong> No heap allocation in op hot paths beyond the output tensor.</li>
<li><strong>N4.</strong> All ops are listed in <code>docs/ops.md</code> with a one-line semantics row, generated<br>
or hand-edited.</li>
</ul>
<hr>
<h2 data-heading="3. Feature Description">3. Feature Description</h2>
<p>The user-visible surface added by this issue:</p>
<pre><code class="language-cpp">// Free functions, in namespace ctorch
Tensor add(const Tensor&#x26; a, const Tensor&#x26; b);
Tensor sub(const Tensor&#x26; a, const Tensor&#x26; b);
Tensor mul(const Tensor&#x26; a, const Tensor&#x26; b);
Tensor div(const Tensor&#x26; a, const Tensor&#x26; b);

void   add_(Tensor&#x26; a, const Tensor&#x26; b);   // in-place
// ... same for sub_, mul_, div_

Tensor neg (const Tensor&#x26; x);
Tensor abs (const Tensor&#x26; x);
Tensor exp (const Tensor&#x26; x);
Tensor log (const Tensor&#x26; x);
Tensor sqrt(const Tensor&#x26; x);
Tensor relu(const Tensor&#x26; x);
Tensor sigmoid(const Tensor&#x26; x);
Tensor tanh(const Tensor&#x26; x);

// Tensor operator overloads
Tensor operator+(const Tensor&#x26;, const Tensor&#x26;);
Tensor operator-(const Tensor&#x26;, const Tensor&#x26;);
Tensor operator*(const Tensor&#x26;, const Tensor&#x26;);
Tensor operator/(const Tensor&#x26;, const Tensor&#x26;);
Tensor operator-(const Tensor&#x26;);             // unary negate
</code></pre>
<p>Example, end-to-end:</p>
<pre><code class="language-cpp">auto a = ctorch::randn({3, 1}).to(Device::cuda(0));
auto b = ctorch::randn({1, 4}).to(Device::cuda(0));
auto c = ctorch::relu(a + b);                // broadcasts to {3, 4}, CUDA kernel
std::cout &#x3C;&#x3C; c.cpu() &#x3C;&#x3C; '\n';
</code></pre>
<hr>
<h2 data-heading="4. Design">4. Design</h2>
<h3 data-heading="4.1 Broadcasting">4.1 Broadcasting</h3>
<p>Implementation lives in <code>src/ops/broadcast.cpp</code>:</p>
<pre><code class="language-cpp">struct BroadcastResult {
  std::vector&#x3C;int64_t> out_shape;
  std::vector&#x3C;int64_t> a_stride;     // padded + virtually broadcast
  std::vector&#x3C;int64_t> b_stride;
};

BroadcastResult broadcast(const Tensor&#x26; a, const Tensor&#x26; b);
</code></pre>
<p>Algorithm (PyTorch/NumPy):</p>
<ol>
<li>Right-align shapes; pad shorter with leading 1s.</li>
<li>For each dimension, sizes must be equal, or one of them <code>1</code>. Any other case →<br>
<code>ctorch::ShapeError</code>.</li>
<li>Output shape takes the max along each dimension.</li>
<li>For an input dimension of size 1 broadcast to N, set its stride to <strong>0</strong> — the same<br>
element is read N times. No copy.</li>
</ol>
<p>The op kernel iterates over <code>out_shape</code> using two stride vectors; this is the same<br>
trick PyTorch's <code>TensorIterator</code> uses.</p>
<h3 data-heading="4.2 Type promotion">4.2 Type promotion</h3>
<p><code>include/ctorch/type_promotion.h</code>:</p>
<pre><code class="language-cpp">dtype promote_types(dtype a, dtype b);   // matches PyTorch's promotion table
</code></pre>
<p>Hardcoded table, no runtime allocation. <code>bool + int32 = int64</code> etc.</p>
<h3 data-heading="4.3 Op registration pattern">4.3 Op registration pattern</h3>
<p>Every op follows this template, located under <code>src/ops/&#x3C;op>.cpp</code>:</p>
<pre><code class="language-cpp">namespace ctorch::op { struct Add; }   // op key — empty tag struct
</code></pre>
<pre><code class="language-cpp">// CPU implementation
namespace {
template &#x3C;class T>
void add_cpu_kernel(const TensorIter&#x26; it) {
  for (auto&#x26; [out, a, b] : it) out = a + b;
}
void add_cpu(const TensorRef&#x26; a, const TensorRef&#x26; b, TensorRef out) {
  DISPATCH_FLOAT_AND_INT(out.dtype(), [&#x26;](auto T) {
    using ScalarT = decltype(T);
    add_cpu_kernel&#x3C;ScalarT>(make_iterator(out, a, b));
  });
}
struct RegisterAddCpu {
  RegisterAddCpu() { dispatch::register_op&#x3C;op::Add>(Device::Kind::CPU, &#x26;add_cpu); }
} _register_add_cpu_;
}  // namespace
</code></pre>
<pre><code class="language-cpp">// CUDA implementation (src/ops/add.cu)
template &#x3C;class T>
__global__ void add_cuda_kernel(IndexerCtx ctx) { ... }
void add_cuda(const TensorRef&#x26; a, const TensorRef&#x26; b, TensorRef out) { ... }
struct RegisterAddCuda { ... } _register_add_cuda_;
</code></pre>
<p>The static initializers run at program start, populating the dispatch table.</p>
<h3 data-heading="4.4 CUDA kernel template">4.4 CUDA kernel template</h3>
<p>One templated kernel handles every element-wise binary op:</p>
<pre><code class="language-cpp">template &#x3C;class T, class Op>
__global__ void elementwise_binary(IndexerCtx ctx, Op op) {
  auto i = blockIdx.x * blockDim.x + threadIdx.x;
  if (i >= ctx.n) return;
  T a = load&#x3C;T>(ctx.a, i);
  T b = load&#x3C;T>(ctx.b, i);
  store&#x3C;T>(ctx.out, i, op(a, b));
}
</code></pre>
<p>Where <code>Op</code> is a small functor (<code>AddOp</code>, <code>MulOp</code>, …). <code>IndexerCtx</code> carries the<br>
broadcast strides; <code>load&#x3C;T>(ptr, i)</code> walks the strides to fetch the element. This keeps<br>
us at one launch per op and avoids per-op kernel duplication.</p>
<p>Launch params: 256 threads/block, grid = <code>ceil_div(N, 256)</code>.</p>
<h3 data-heading="4.5 CPU kernel template">4.5 CPU kernel template</h3>
<p>Two paths:</p>
<ul>
<li><strong>Contiguous fast path</strong> (both inputs and output contiguous, same dtype): tight<br>
<code>for (i = 0; i &#x3C; n; ++i) out[i] = op(a[i], b[i])</code>. Compiler auto-vectorizes; we add<br>
<code>#pragma omp simd</code> as a hint.</li>
<li><strong>General path</strong>: stride-based iterator (same indexer as CUDA, but in host code).</li>
</ul>
<p>The iterator is shared between paths via <code>make_iterator(out, a, b)</code> returning a<br>
zip-like view.</p>
<h3 data-heading="4.6 In-place variants">4.6 In-place variants</h3>
<p><code>add_(t, u)</code> checks:</p>
<ul>
<li>Output (i.e. <code>t</code>) shape must equal the broadcast result shape (no implicit reshape).</li>
<li><code>t</code> and <code>u</code> must not alias unless they are the exact same <code>Storage</code> and <code>offset</code>.<br>
Otherwise → throw.</li>
</ul>
<h3 data-heading="4.7 Files added by this issue">4.7 Files added by this issue</h3>
<pre><code>include/ctorch/
├── ops/
│   ├── elementwise.h        # public function declarations
│   └── op_keys.h            # tag structs (Add, Sub, …)
└── type_promotion.h

src/
├── ops/
│   ├── broadcast.cpp / .h
│   ├── tensor_iter.cpp / .h
│   ├── add.cpp              # CPU
│   ├── add.cu               # CUDA
│   ├── sub.cpp / .cu
│   ├── mul.cpp / .cu
│   ├── div.cpp / .cu
│   └── unary.{cpp,cu}       # all unary ops in one TU pair
└── type_promotion.cpp

tests/
├── ops/
│   ├── add_parity_test.cpp
│   ├── broadcast_test.cpp
│   ├── promotion_test.cpp
│   └── unary_parity_test.cpp
├── parity/
│   ├── fixtures/            # *.npy files, regenerated by scripts/gen_parity.py
│   └── load_npy.{h,cpp}
└── bench/
    └── add_bench.cpp        # microbenchmark vs hand-written kernel
</code></pre>
<h3 data-heading="4.8 Parity fixture pipeline">4.8 Parity fixture pipeline</h3>
<p><code>scripts/gen_parity.py</code> (run once on the developer's machine, output committed):</p>
<pre><code class="language-python">import torch, numpy as np
torch.manual_seed(0)
for shape_a, shape_b in [((3, 4), (3, 4)), ((3, 1), (1, 4)), ...]:
    a = torch.randn(shape_a); b = torch.randn(shape_b)
    np.save(f"add_{shape_a}_{shape_b}.npy", (a + b).numpy())
    ...
</code></pre>
<p>C++ tests load the <code>.npy</code> files via a tiny header-only loader (<code>load_npy.h</code>) and assert<br>
<code>max(|out - ref|) &#x3C;= tol</code> on each fixture.</p>
<p>This decouples the C++ test suite from a Python runtime — CI doesn't need PyTorch<br>
installed.</p>
<hr>
<h2 data-heading="5. Acceptance Criteria">5. Acceptance Criteria</h2>
<ul class="contains-task-list">
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">All ops in §3 are callable on both CPU and CUDA tensors and select the correct<br>
kernel via dispatch.</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">Broadcasting unit test covers: scalar+vector, <code>(3,1)+(1,4)</code>, <code>(2,3,4)+(3,4)</code>,<br>
and confirms zero-stride trick (no extra memory).</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">Parity tests pass for every op against fixed <code>.npy</code> fixtures with the tolerances<br>
in N1.</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">Type promotion test: each pair in <code>{float32, float64, int32, int64, bool}</code> x<br>
<code>{float32, float64, int32, int64, bool}</code> agrees with PyTorch's <code>torch.result_type</code>.</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">Microbenchmark: CUDA <code>add</code> on <code>1 &#x3C;&#x3C; 20</code> fp32 elements runs within 10 % of the<br>
reference kernel. Result reported in CI logs.</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">In-place op throws <code>AliasError</code> when output aliases a non-trivial slice of an<br>
input.</li>
<li class="task-list-item" data-task=" "><input class="task-list-item-checkbox" type="checkbox">No additional heap allocation (besides the output tensor) is observed in the<br>
CPU <code>add</code> path under a sampling allocator hook.</li>
</ul>
<hr>
<h2 data-heading="6. Suggested commit breakdown">6. Suggested commit breakdown</h2>
<ol>
<li><code>feat(ops): broadcast computation and tensor iterator</code></li>
<li><code>feat(ops): type promotion table</code></li>
<li><code>feat(ops): CPU + CUDA add/sub/mul/div with parity tests</code></li>
<li><code>feat(ops): unary ops (neg/abs/exp/log/sqrt/relu/sigmoid/tanh)</code></li>
<li><code>feat(ops): in-place variants and alias detection</code></li>
<li><code>chore(test): commit npy parity fixtures and gen script</code></li>
<li><code>bench: add CUDA add microbenchmark vs reference kernel</code></li>
</ol>
<hr>
<h2 data-heading="7. Risks and trade-offs">7. Risks and trade-offs</h2>

Risk | Mitigation
-- | --
.npy loader brings encoding bugs | Vendor a tested 200-line loader; restrict to float32/float64/int32/int64
CUDA template instantiation explosion | Single templated indexer kernel, op functor injected; nvcc compile time monitored in CI
Floating-point parity fails on exp/log due to libm vs cuMath differences | Per-op tolerance table, documented in docs/ops.md; use 1e-4 rel for transcendentals
In-place alias detection is brittle | Conservative check: same Storage* + overlapping offsets → throw. False positives are acceptable
Auto-vectorization regresses on certain compilers | Microbenchmark pinned in CI; regression threshold = 1.2x slower fails the build

<hr>
<h2 data-heading="8. Out of scope">8. Out of scope</h2>
<ul>
<li>Reductions (<code>sum</code>, <code>mean</code>, <code>max</code>) — separate <code>[Feature]</code> issue.</li>
<li>Indexing / slicing — separate <code>[Feature]</code> issue.</li>
<li>Matmul / transpose / linear algebra (cuBLAS) — separate <code>[Feature]</code> issue.</li>
<li>Backward / autograd integration — Autograd milestone.</li>
<li><code>bfloat16</code> / <code>float16</code> arithmetic — Mixed-precision milestone.</li>
<li>Operator fusion — Performance milestone.</li>
</ul>
